File: windows/src/LanStash.Infrastructure/DsmQuickConnectResolver.cs
using System.Security.Cryptography;
using System.Text;
using System.Text.Json;
using System.Text.Json.Nodes;
using LanStash.Domain;

namespace LanStash.Infrastructure;

public enum QuickConnectEndpointKind
{
    Local,
    External,
    Relay,
}

public sealed record QuickConnectEndpoint(string Host, int Port, QuickConnectEndpointKind Kind);

/// <summary>
/// 受限使用 Synology QuickConnect 当前客户端契约。
/// get_server_info、request_tunnel 属于未公开的内部 API；登录前会限制官方域名并验证中继身份。
/// </summary>
public sealed class DsmQuickConnectResolver(HttpClient httpClient)
{
    private const int MaximumResponseBytes = 1024 * 1024;
    private readonly HttpClient _http = httpClient;
    private static readonly string[] ControlUrls =
    [
        "https://global.quickconnect.to/Serv.php",
        "https://global.quickconnect.cn/Serv.php",
    ];

    public async Task<IReadOnlyList<QuickConnectEndpoint>> ResolveAsync(
        string id,
        CancellationToken cancellationToken = default)
    {
        ValidateId(id);
        DsmException? lastError = null;
        foreach (var controlUrl in ControlUrls)
        {
            try
            {
                return DecodeEndpoints(await SendAsync(
                    "get_server_info", false, id, controlUrl, cancellationToken).ConfigureAwait(false));
            }
            catch (DsmException error)
            {
                lastError = error;
            }
        }
        throw lastError ?? ServiceUnavailable();
    }

    public async Task<QuickConnectEndpoint> RequestRelayAsync(
        string id,
        CancellationToken cancellationToken = default)
    {
        ValidateId(id);
        var controlUrl = await ResolveControlUrlAsync(id, cancellationToken).ConfigureAwait(false);
        DsmException? lastError = null;
        for (var attempt = 0; attempt < 3; attempt++)
        {
            try
            {
                var descriptor = DecodeRelayDescriptor(await SendAsync(
                    "request_tunnel", true, id, controlUrl, cancellationToken).ConfigureAwait(false), id);
                await VerifyRelayAsync(descriptor, cancellationToken).ConfigureAwait(false);
                return descriptor.Endpoint;
            }
            catch (DsmException error) when (
                error.Message is not "这台 NAS 没有开启 QuickConnect 中继" and
                not "QuickConnect 返回的连接无法确认属于这台 NAS")
            {
                lastError = error;
                if (attempt < 2)
                {
                    await Task.Delay((attempt + 1) * 1000, cancellationToken).ConfigureAwait(false);
                }
            }
        }
        throw lastError ?? RelayUnavailable();
    }

    public static IReadOnlyList<QuickConnectEndpoint> DecodeEndpoints(JsonArray responses)
    {
        var response = SuccessfulResponse(responses);
        if (!string.Equals(response["server"]?["ds_state"]?.GetValue<string>(), "CONNECTED",
                StringComparison.OrdinalIgnoreCase))
        {
            throw new DsmException(
                "QuickConnect 找到了这台 NAS，但设备目前不在线",
                "确认 NAS 已开机并联网后重试。");
        }
        var port = response["service"]?["port"]?.GetValue<int>() ?? 0;
        var smartDns = response["smartdns"] as JsonObject;
        if (port is < 1 or > 65535 || smartDns is null)
        {
            throw NoDirectRoute();
        }

        var endpoints = new List<QuickConnectEndpoint>();
        var seen = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
        if (smartDns["lan"] is JsonArray lan)
        {
            foreach (var value in lan)
            {
                var host = value is JsonObject item
                    ? item["host"]?.GetValue<string>()
                    : value?.GetValue<string>();
                AddTrustedDirect(endpoints, seen, host, port, QuickConnectEndpointKind.Local);
            }
        }
        AddTrustedDirect(
            endpoints,
            seen,
            smartDns["host"]?.GetValue<string>(),
            port,
            QuickConnectEndpointKind.External);
        return endpoints.Count > 0 ? endpoints : throw NoDirectRoute();
    }

    private async Task<string> ResolveControlUrlAsync(string id, CancellationToken cancellationToken)
    {
        DsmException? lastError = null;
        foreach (var controlUrl in ControlUrls)
        {
            try
            {
                var response = SuccessfulResponse(await SendAsync(
                    "get_server_info", false, id, controlUrl, cancellationToken).ConfigureAwait(false));
                var host = response["env"]?["control_host"]?.GetValue<string>()?.ToLowerInvariant();
                if (host is null || !IsTrustedControlHost(host))
                {
                    throw InvalidResponse();
                }
                return $"https://{host}/Serv.php";
            }
            catch (DsmException error)
            {
                lastError = error;
            }
        }
        throw lastError ?? ServiceUnavailable();
    }

    private async Task<JsonArray> SendAsync(
        string command,
        bool stopWhenSuccess,
        string serverId,
        string controlUrl,
        CancellationToken cancellationToken)
    {
        var payload = new JsonArray
        {
            new JsonObject
            {
                ["version"] = 1,
                ["command"] = command,
                ["stop_when_error"] = false,
                ["stop_when_success"] = stopWhenSuccess,
                ["id"] = "mainapp_https",
                ["serverID"] = serverId,
                ["is_gofile"] = false,
                ["path"] = "",
            },
        };
        using var request = new HttpRequestMessage(HttpMethod.Post, controlUrl)
        {
            Content = new StringContent(payload.ToJsonString(), Encoding.UTF8, "application/json"),
        };
        request.Headers.Accept.ParseAdd("application/json");
        request.Headers.UserAgent.ParseAdd("LanStash-Windows/0.1");
        using var timeout = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken);
        timeout.CancelAfter(command == "request_tunnel" ? TimeSpan.FromSeconds(30) : TimeSpan.FromSeconds(15));
        try
        {
            using var response = await _http.SendAsync(
                request, HttpCompletionOption.ResponseHeadersRead, timeout.Token).ConfigureAwait(false);
            if (!response.IsSuccessStatusCode)
            {
                throw ServiceUnavailable();
            }
            return await ReadJsonArrayAsync(response, timeout.Token).ConfigureAwait(false);
        }
        catch (OperationCanceledException) when (!cancellationToken.IsCancellationRequested)
        {
            throw ServiceUnavailable();
        }
        catch (HttpRequestException)
        {
            throw ServiceUnavailable();
        }
    }

    private async Task VerifyRelayAsync(RelayDescriptor descriptor, CancellationToken cancellationToken)
    {
        if (!descriptor.PingPongPath.StartsWith('/') ||
            descriptor.PingPongPath.Length > 2048 ||
            descriptor.PingPongPath.Contains("://", StringComparison.Ordinal) ||
            descriptor.PingPongPath.Contains('#'))
        {
            throw InvalidResponse();
        }
        var expectedId = Convert.ToHexString(
            MD5.HashData(Encoding.UTF8.GetBytes(descriptor.ServerId))).ToLowerInvariant();
        var uri = new Uri($"https://{descriptor.Endpoint.Host}{descriptor.PingPongPath}");
        for (var attempt = 0; attempt < 6; attempt++)
        {
            try
            {
                using var request = new HttpRequestMessage(HttpMethod.Get, uri);
                request.Headers.Accept.ParseAdd("application/json");
                request.Headers.UserAgent.ParseAdd("LanStash-Windows/0.1");
                using var response = await _http.SendAsync(
                    request, HttpCompletionOption.ResponseHeadersRead, cancellationToken).ConfigureAwait(false);
                if (!response.IsSuccessStatusCode)
                {
                    throw RelayUnavailable();
                }
                var json = await ReadJsonObjectAsync(response, cancellationToken).ConfigureAwait(false);
                if (!string.Equals(json["ezid"]?.GetValue<string>(), expectedId, StringComparison.OrdinalIgnoreCase))
                {
                    throw new DsmException(
                        "QuickConnect 返回的连接无法确认属于这台 NAS",
                        "为保护登录信息，岚仓已停止连接。");
                }
                return;
            }
            catch (DsmException error) when (
                error.Message != "QuickConnect 返回的连接无法确认属于这台 NAS" && attempt < 5)
            {
                await Task.Delay(1000, cancellationToken).ConfigureAwait(false);
            }
            catch (HttpRequestException) when (attempt < 5)
            {
                await Task.Delay(1000, cancellationToken).ConfigureAwait(false);
            }
        }
        throw RelayUnavailable();
    }

    private static RelayDescriptor DecodeRelayDescriptor(JsonArray responses, string id)
    {
        if (responses.OfType<JsonObject>().Any(item => item["errno"]?.GetValue<int>() == 19))
        {
            throw new DsmException(
                "这台 NAS 没有开启 QuickConnect 中继",
                "请在 DSM 的 QuickConnect 高级设置中开启中继后重试。");
        }
        var response = SuccessfulResponse(responses);
        var service = response["service"] as JsonObject;
        var server = response["server"] as JsonObject;
        var environment = response["env"] as JsonObject;
        var relayPort = service?["relay_port"]?.GetValue<int>() ?? 0;
        var serverId = server?["serverID"]?.GetValue<string>();
        var region = environment?["relay_region"]?.GetValue<string>()?.ToLowerInvariant();
        var controlHost = environment?["control_host"]?.GetValue<string>()?.ToLowerInvariant();
        if (string.IsNullOrWhiteSpace(service?["relay_ip"]?.GetValue<string>()) ||
            relayPort is < 1 or > 65535 ||
            string.IsNullOrWhiteSpace(serverId) ||
            region is null || !IsValidHostLabel(region) ||
            controlHost is null || !IsTrustedControlHost(controlHost))
        {
            throw RelayUnavailable();
        }
        var topDomain = controlHost.EndsWith(".quickconnect.cn", StringComparison.Ordinal) ? "cn" : "to";
        var host = $"{id.ToLowerInvariant()}.{region}.quickconnect.{topDomain}";
        if (!IsTrustedRelayHost(host))
        {
            throw InvalidResponse();
        }
        return new(
            new QuickConnectEndpoint(host, 443, QuickConnectEndpointKind.Relay),
            serverId,
            server?["pingpong_path"]?.GetValue<string>()
                ?? "/webman/pingpong.cgi?action=cors&quickconnect=true");
    }

    public static bool IsTrustedRelayHost(string host)
    {
        var labels = host.ToLowerInvariant().Split('.');
        return labels.Length == 4 &&
            labels[2] == "quickconnect" &&
            labels[3] is "to" or "cn" &&
            IsValidHostLabel(labels[0]) &&
            IsValidHostLabel(labels[1]);
    }

    private static JsonObject SuccessfulResponse(JsonArray responses) =>
        responses.OfType<JsonObject>().FirstOrDefault(item => item["errno"]?.GetValue<int>() == 0)
        ?? throw new DsmException(
            "没有找到这个 QuickConnect ID",
            "请检查拼写和 NAS 中的 QuickConnect 设置。");

    private static void AddTrustedDirect(
        ICollection<QuickConnectEndpoint> endpoints,
        ISet<string> seen,
        string? rawHost,
        int port,
        QuickConnectEndpointKind kind)
    {
        var host = rawHost?.ToLowerInvariant();
        if (host is not null && IsTrustedDirectHost(host) && seen.Add(host))
        {
            endpoints.Add(new(host, port, kind));
        }
    }

    private static bool IsTrustedDirectHost(string host)
    {
        var labels = host.Split('.');
        return labels.Length >= 4 &&
            labels[^3] == "direct" &&
            labels[^2] == "quickconnect" &&
            labels[^1] is "to" or "cn" &&
            labels.All(IsValidHostLabel);
    }

    private static bool IsTrustedControlHost(string host) =>
        (host.EndsWith(".quickconnect.to", StringComparison.Ordinal) ||
         host.EndsWith(".quickconnect.cn", StringComparison.Ordinal)) &&
        host.Split('.').All(IsValidHostLabel);

    private static bool IsValidHostLabel(string value) =>
        value.Length is >= 1 and <= 63 &&
        value[0] != '-' &&
        value[^1] != '-' &&
        value.All(character =>
            character is >= 'a' and <= 'z' or >= 'A' and <= 'Z' or >= '0' and <= '9' or '-');

    private static void ValidateId(string id)
    {
        if (!NasAddressParser.IsPotentialQuickConnectId(id))
        {
            throw new DsmException("无法识别这个 QuickConnect ID", "请检查拼写后重试。");
        }
    }

    private static async Task<JsonArray> ReadJsonArrayAsync(
        HttpResponseMessage response,
        CancellationToken cancellationToken)
    {
        var node = await ReadJsonAsync(response, cancellationToken).ConfigureAwait(false);
        return node as JsonArray ?? throw InvalidResponse();
    }

    private static async Task<JsonObject> ReadJsonObjectAsync(
        HttpResponseMessage response,
        CancellationToken cancellationToken)
    {
        var node = await ReadJsonAsync(response, cancellationToken).ConfigureAwait(false);
        return node as JsonObject ?? throw RelayUnavailable();
    }

    private static async Task<JsonNode?> ReadJsonAsync(
        HttpResponseMessage response,
        CancellationToken cancellationToken)
    {
        if (response.Content.Headers.ContentLength > MaximumResponseBytes)
        {
            throw InvalidResponse();
        }
        await using var source = await response.Content.ReadAsStreamAsync(cancellationToken).ConfigureAwait(false);
        using var memory = new MemoryStream();
        var buffer = new byte[8192];
        while (true)
        {
            var count = await source.ReadAsync(buffer, cancellationToken).ConfigureAwait(false);
            if (count == 0)
            {
                break;
            }
            if (memory.Length + count > MaximumResponseBytes)
            {
                throw InvalidResponse();
            }
            await memory.WriteAsync(buffer.AsMemory(0, count), cancellationToken).ConfigureAwait(false);
        }
        memory.Position = 0;
        return await JsonNode.ParseAsync(memory, cancellationToken: cancellationToken).ConfigureAwait(false);
    }

    private static DsmException NoDirectRoute() => new(
        "QuickConnect 没有提供可用的直接连接",
        "岚仓将继续尝试安全中继。");
    private static DsmException ServiceUnavailable() => new(
        "QuickConnect 暂时没有响应",
        "请稍后重试。");
    private static DsmException InvalidResponse() => new(
        "QuickConnect 返回的信息无法读取",
        "请稍后重试。");
    private static DsmException RelayUnavailable() => new(
        "QuickConnect 暂时无法建立中继连接",
        "请稍后重试。");

    private sealed record RelayDescriptor(
        QuickConnectEndpoint Endpoint,
        string ServerId,
        string PingPongPath);
}

File: android/app/src/main/java/io/github/qwertyuiop1995/dsmnativeclient/network/DsmQuickConnectResolver.kt
package io.github.qwertyuiop1995.dsmnativeclient.network

import io.github.qwertyuiop1995.dsmnativeclient.domain.DsmFailure
import java.io.IOException
import java.security.MessageDigest
import java.util.Locale
import kotlin.coroutines.resume
import kotlin.coroutines.resumeWithException
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.TimeoutCancellationException
import kotlinx.coroutines.delay
import kotlinx.coroutines.suspendCancellableCoroutine
import kotlinx.coroutines.withTimeout
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.buildJsonArray
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.put
import okhttp3.Call
import okhttp3.Callback
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import okhttp3.Response
import okio.Buffer

internal enum class QuickConnectEndpointKind {
    LOCAL,
    EXTERNAL,
    RELAY,
}

internal data class QuickConnectEndpoint(
    val host: String,
    val port: Int,
    val kind: QuickConnectEndpointKind,
)

internal data class QuickConnectRelayDescriptor(
    val endpoint: QuickConnectEndpoint,
    val serverId: String,
    val pingPongPath: String,
)

/**
 * QuickConnect 的受限解析器。
 *
 * `get_server_info` 与 `request_tunnel` 来自 Synology QuickConnect 当前客户端契约，
 * 属于未公开的内部 API。发送登录信息前必须限制官方域名并校验中继 NAS 身份。
 */
internal class DsmQuickConnectResolver(
    private val http: OkHttpClient = OkHttpClient.Builder()
        .followRedirects(false)
        .followSslRedirects(false)
        .retryOnConnectionFailure(true)
        .build(),
    private val json: Json = Json { ignoreUnknownKeys = true },
) {
    private val controlUrls: List<String> = if (
        Locale.getDefault().country.equals("CN", ignoreCase = true)
    ) {
        listOf(CHINA_CONTROL_URL, GLOBAL_CONTROL_URL)
    } else {
        listOf(GLOBAL_CONTROL_URL, CHINA_CONTROL_URL)
    }

    suspend fun resolve(id: String): List<QuickConnectEndpoint> {
        validateId(id)
        var lastFailure = serviceUnavailable()
        for (controlUrl in controlUrls) {
            try {
                return decodeEndpoints(
                    send("get_server_info", stopWhenSuccess = false, id, controlUrl)
                )
            } catch (error: CancellationException) {
                throw error
            } catch (error: DsmFailure) {
                lastFailure = error
            }
        }
        throw lastFailure
    }

    suspend fun requestRelay(id: String): QuickConnectEndpoint {
        validateId(id)
        val controlUrl = resolveControlUrl(id)
        var lastFailure = relayUnavailable()
        repeat(3) { attempt ->
            try {
                val descriptor = decodeRelayDescriptor(
                    send("request_tunnel", stopWhenSuccess = true, id, controlUrl),
                    id,
                )
                verifyRelay(descriptor)
                return descriptor.endpoint
            } catch (error: CancellationException) {
                throw error
            } catch (error: DsmFailure) {
                if (error.message == RELAY_DISABLED_MESSAGE ||
                    error.message == RELAY_IDENTITY_MISMATCH_MESSAGE
                ) {
                    throw error
                }
                lastFailure = error
                if (attempt < 2) delay((attempt + 1) * 1_000L)
            }
        }
        throw lastFailure
    }

    private suspend fun resolveControlUrl(id: String): String {
        var lastFailure = serviceUnavailable()
        for (controlUrl in controlUrls) {
            try {
                val host = decodeControlHost(
                    send("get_server_info", stopWhenSuccess = false, id, controlUrl)
                )
                return "https://$host/Serv.php"
            } catch (error: CancellationException) {
                throw error
            } catch (error: DsmFailure) {
                lastFailure = error
            }
        }
        throw lastFailure
    }

    private suspend fun send(
        command: String,
        stopWhenSuccess: Boolean,
        serverId: String,
        controlUrl: String,
    ): JsonArray {
        val body = buildJsonArray {
            add(
                buildJsonObject {
                    put("version", 1)
                    put("command", command)
                    put("stop_when_error", false)
                    put("stop_when_success", stopWhenSuccess)
                    put("id", "mainapp_https")
                    put("serverID", serverId)
                    put("is_gofile", false)
                    put("path", "")
                }
            )
        }.toString().toRequestBody(JSON_MEDIA_TYPE)
        val request = Request.Builder()
            .url(controlUrl)
            .post(body)
            .header("Accept", "application/json")
            .header("User-Agent", "LanStash-Android/0.1")
            .build()
        val timeout = if (command == "request_tunnel") 30_000L else 15_000L
        return try {
            withTimeout(timeout) {
                http.newCall(request).await().use { response ->
                    if (!response.isSuccessful) throw serviceUnavailable()
                    val bytes = response.readBoundedBody(::invalidResponse)
                    runCatching { json.parseToJsonElement(bytes.decodeToString()) as JsonArray }
                        .getOrElse { throw invalidResponse() }
                }
            }
        } catch (_: TimeoutCancellationException) {
            throw serviceUnavailable()
        }
    }

    private suspend fun verifyRelay(descriptor: QuickConnectRelayDescriptor) {
        val path = descriptor.pingPongPath
        if (!path.startsWith('/') ||
            path.length > 2_048 ||
            "://" in path ||
            '#' in path
        ) {
            throw invalidResponse()
        }
        val expectedId = MessageDigest.getInstance("MD5")
            .digest(descriptor.serverId.toByteArray(Charsets.UTF_8))
            .joinToString("") { byte -> "%02x".format(byte.toInt() and 0xff) }
        val url = "https://${descriptor.endpoint.host}$path"

        repeat(6) { attempt ->
            try {
                val request = Request.Builder()
                    .url(url)
                    .get()
                    .header("Accept", "application/json")
                    .header("User-Agent", "LanStash-Android/0.1")
                    .build()
                val responseJson = try {
                    withTimeout(15_000L) {
                        http.newCall(request).await().use { response ->
                            if (!response.isSuccessful) throw relayUnavailable()
                            val bytes = response.readBoundedBody(::relayUnavailable)
                            runCatching {
                                json.parseToJsonElement(bytes.decodeToString()).jsonObject
                            }.getOrElse { throw relayUnavailable() }
                        }
                    }
                } catch (_: TimeoutCancellationException) {
                    throw relayUnavailable()
                }
                val actualId = responseJson.string("ezid")?.lowercase()
                if (actualId != expectedId) throw relayIdentityMismatch()
                return
            } catch (error: CancellationException) {
                throw error
            } catch (error: DsmFailure) {
                if (error.message == RELAY_IDENTITY_MISMATCH_MESSAGE) throw error
                if (attempt < 5) delay(1_000L)
            }
        }
        throw relayUnavailable()
    }

    internal fun decodeEndpoints(responses: JsonArray): List<QuickConnectEndpoint> {
        val response = successfulResponse(responses)
        if (response.objectValue("server")?.string("ds_state")?.uppercase() != "CONNECTED") {
            throw DsmFailure(
                null,
                "QuickConnect 找到了这台 NAS，但设备目前不在线",
                "确认 NAS 已开机并联网后重试。",
            )
        }
        val service = response.objectValue("service") ?: throw noDirectRoute()
        val port = service.int("port")?.takeIf { it in 1..65_535 } ?: throw noDirectRoute()
        val smartDns = response.objectValue("smartdns") ?: throw noDirectRoute()
        val seen = mutableSetOf<String>()
        val local = smartDns.arrayValue("lan").mapNotNull { element ->
            val host = runCatching { element.jsonObject }.getOrNull()?.string("host")
                ?: runCatching { element.toString().trim('"') }.getOrNull()
            trustedDirectEndpoint(host, port, QuickConnectEndpointKind.LOCAL, seen)
        }
        val external = listOfNotNull(smartDns.string("host")).mapNotNull { host ->
            trustedDirectEndpoint(host, port, QuickConnectEndpointKind.EXTERNAL, seen)
        }
        return (local + external).ifEmpty { throw noDirectRoute() }
    }

    internal fun decodeControlHost(responses: JsonArray): String {
        val response = successfulResponse(responses)
        if (response.objectValue("server")?.string("ds_state")?.uppercase() != "CONNECTED") {
            throw invalidResponse()
        }
        val host = response.objectValue("env")?.string("control_host")?.lowercase()
            ?: throw invalidResponse()
        if (!isTrustedControlHost(host)) throw invalidResponse()
        return host
    }

    internal fun decodeRelayDescriptor(
        responses: JsonArray,
        quickConnectId: String,
    ): QuickConnectRelayDescriptor {
        if (responses.any { element ->
                runCatching { element.jsonObject.int("errno") == 19 }.getOrDefault(false)
            }
        ) {
            throw DsmFailure(
                null,
                RELAY_DISABLED_MESSAGE,
                "请在 DSM 的 QuickConnect 高级设置中开启中继后重试。",
            )
        }
        val response = successfulResponse(responses)
        val service = response.objectValue("service") ?: throw relayUnavailable()
        if (service.string("relay_ip").isNullOrBlank()) throw relayUnavailable()
        service.int("relay_port")?.takeIf { it in 1..65_535 } ?: throw relayUnavailable()
        val server = response.objectValue("server") ?: throw relayUnavailable()
        val serverId = server.string("serverID")?.takeIf(String::isNotBlank)
            ?: throw relayUnavailable()
        val environment = response.objectValue("env") ?: throw relayUnavailable()
        val region = environment.string("relay_region")?.lowercase()
            ?.takeIf(::isValidHostLabel)
            ?: throw relayUnavailable()
        val controlHost = environment.string("control_host")?.lowercase()
            ?.takeIf(::isTrustedControlHost)
            ?: throw relayUnavailable()
        val topDomain = if (controlHost.endsWith(".quickconnect.cn")) "cn" else "to"
        val host = "${quickConnectId.lowercase()}.$region.quickconnect.$topDomain"
        if (!isTrustedRelayHost(host)) throw invalidResponse()
        return QuickConnectRelayDescriptor(
            endpoint = QuickConnectEndpoint(host, 443, QuickConnectEndpointKind.RELAY),
            serverId = serverId,
            pingPongPath = server.string("pingpong_path")
                ?: "/webman/pingpong.cgi?action=cors&quickconnect=true",
        )
    }

    private fun successfulResponse(responses: JsonArray): JsonObject =
        responses.mapNotNull { runCatching { it.jsonObject }.getOrNull() }
            .firstOrNull { it.int("errno") == 0 }
            ?: throw DsmFailure(
                null,
                "没有找到这个 QuickConnect ID",
                "请检查拼写和 NAS 中的 QuickConnect 设置。",
            )

    private fun trustedDirectEndpoint(
        rawHost: String?,
        port: Int,
        kind: QuickConnectEndpointKind,
        seen: MutableSet<String>,
    ): QuickConnectEndpoint? {
        val host = rawHost?.lowercase()?.takeIf(::isTrustedDirectHost) ?: return null
        if (!seen.add(host)) return null
        return QuickConnectEndpoint(host, port, kind)
    }

    private fun validateId(id: String) {
        if (!NasAddressParser.isPotentialQuickConnectId(id)) {
            throw DsmFailure(
                null,
                "无法识别这个 QuickConnect ID",
                "请检查拼写后重试。",
            )
        }
    }

    private fun isTrustedDirectHost(host: String): Boolean {
        val labels = host.split('.')
        return labels.size >= 4 &&
            labels[labels.lastIndex - 2] == "direct" &&
            labels[labels.lastIndex - 1] == "quickconnect" &&
            labels.last() in setOf("to", "cn") &&
            labels.all(::isValidHostLabel)
    }

    internal fun isTrustedRelayHost(host: String): Boolean {
        val labels = host.lowercase().split('.')
        return labels.size == 4 &&
            labels[2] == "quickconnect" &&
            labels[3] in setOf("to", "cn") &&
            isValidHostLabel(labels[0]) &&
            isValidHostLabel(labels[1])
    }

    private fun isTrustedControlHost(host: String): Boolean =
        (host.endsWith(".quickconnect.to") || host.endsWith(".quickconnect.cn")) &&
            host.split('.').all(::isValidHostLabel)

    private fun isValidHostLabel(value: String): Boolean =
        value.length in 1..63 &&
            !value.startsWith('-') &&
            !value.endsWith('-') &&
            value.all { character ->
                character in 'a'..'z' ||
                    character in 'A'..'Z' ||
                    character in '0'..'9' ||
                    character == '-'
            }

    private fun noDirectRoute() = DsmFailure(
        null,
        "QuickConnect 没有提供可用的直接连接",
        "岚仓将继续尝试安全中继。",
    )

    private fun serviceUnavailable() = DsmFailure(
        null,
        "QuickConnect 暂时没有响应",
        "请稍后重试。",
    )

    private fun invalidResponse() = DsmFailure(
        null,
        "QuickConnect 返回的信息无法读取",
        "请稍后重试。",
    )

    private fun relayUnavailable() = DsmFailure(
        null,
        "QuickConnect 暂时无法建立中继连接",
        "请稍后重试。",
    )

    private fun relayIdentityMismatch() = DsmFailure(
        null,
        RELAY_IDENTITY_MISMATCH_MESSAGE,
        "为保护登录信息，岚仓已停止连接。",
    )

    private fun Response.readBoundedBody(failure: () -> DsmFailure): ByteArray {
        val responseBody = body ?: return byteArrayOf()
        val contentLength = responseBody.contentLength()
        if (contentLength > MAXIMUM_RESPONSE_BYTES) throw failure()

        val source = responseBody.source()
        val buffer = Buffer()
        var totalBytes = 0L
        while (true) {
            val remaining = MAXIMUM_RESPONSE_BYTES + 1L - totalBytes
            if (remaining <= 0L) throw failure()
            val read = source.read(buffer, minOf(8_192L, remaining))
            if (read == -1L) break
            totalBytes += read
        }
        if (totalBytes > MAXIMUM_RESPONSE_BYTES) throw failure()
        return buffer.readByteArray()
    }

    private suspend fun Call.await(): Response = suspendCancellableCoroutine { continuation ->
        continuation.invokeOnCancellation { cancel() }
        enqueue(object : Callback {
            override fun onFailure(call: Call, error: IOException) {
                if (!continuation.isCancelled) {
                    continuation.resumeWithException(serviceUnavailable())
                }
            }

            override fun onResponse(call: Call, response: Response) {
                continuation.resume(response)
            }
        })
    }

    private companion object {
        const val MAXIMUM_RESPONSE_BYTES = 1_024 * 1_024L
        const val GLOBAL_CONTROL_URL = "https://global.quickconnect.to/Serv.php"
        const val CHINA_CONTROL_URL = "https://global.quickconnect.cn/Serv.php"
        const val RELAY_DISABLED_MESSAGE = "这台 NAS 没有开启 QuickConnect 中继"
        const val RELAY_IDENTITY_MISMATCH_MESSAGE = "QuickConnect 返回的连接无法确认属于这台 NAS"
        val JSON_MEDIA_TYPE = "application/json".toMediaType()
    }
}
